src/blockMonitor/blockCache.ts

```diff
@@ -0,0 +1,230 @@
+        for (let txHash of transactionHashes) {
+            removeItemFromKeyedSet(this.blockHashesByTxHash, txHash, blockHash);
+        }
+
+        // Remove stored set of transactions for this block
+        this.txHashesByBlockHash.delete(blockHash);
+    }
+
+    // Remove all the blocks that are deeper than maxDepth, and all connected information.
+    private prune() {
+        for (let height = this.minStoredHeight; height < this.getMinVisibleHeight(); height++) {
+            const hashesByHeight = this.blockHashesByHeight.get(height);
+            if (hashesByHeight !== undefined) {
+                for (let hash of hashesByHeight) {
+                    this.pruneBlock(hash);
+                }
+                this.blockHashesByHeight.delete(height);
+            }
+        }
+        this.minStoredHeight = this.mMaxHeight - this.maxDepth;
+    }
+
+    // Makes a new block stub, linking the parent if available
+    private makeBlockStub(hash: string, number: number, parentHash: string) {
+        const parentBlockStubChain = this.blockStubsByHash.get(parentHash);
+        let newBlockStubChain: BlockStubChain;
+        if (parentBlockStubChain === undefined) {
+            newBlockStubChain = BlockStubChain.newRoot({ hash, number, parentHash });
+        } else {
+            newBlockStubChain = parentBlockStubChain.extend({ hash, number, parentHash });
+        }
+        return newBlockStubChain;
+    }
+
+    /**
+     * Returns true it `block` can be added to the cache, that is if either:
+     *   - it is the first block ever seen, or
+     *   - its parent is already in the cache, or
+     *   - it is at depth least `this.maxDepth`.
+     * @param block
+     */
+    public canAddBlock(block: ethers.providers.Block): boolean {
+        return (
+            this.initialHeight === -1 || this.hasBlock(block.parentHash) || block.number <= this.getMinVisibleHeight()
+        );
+    }
+
+    /**
+     * Adds `block`to the cache.
+     * @param block
+     * @returns `true` if the block was added, `false` if the block was not added because too deep or already there.
+     * @throws `ApplicationError` if the block cannot be added because its parent is not in cache.
+     */
+    public addBlock(block: ethers.providers.Block): boolean {
+        if (this.initialHeight === -1) {
+            // First block added, blocks before this point will not be stored.#
+            this.initialHeight = block.number;
+        } else {
+            if (this.blockStubsByHash.has(block.hash)) {
+                // block already in memory
+                return false;
+            }
+
+            if (block.number < this.getMinVisibleHeight()) {
+                // block too deep
+                return false;
+            }
+
+            // If the block's parent is above the minimum visible height, it needs to be added first
+            if (!this.canAddBlock(block)) {
```

Comment:
I wanted to distinguish the situation where you try to add a block that is not necessary (i.e. would be pruned immediately, or block already added), from syntactically illegal attempts (parent should be added first). So I can return `false` for legal-but-redundant, and throw an exception for illegal attempts.
Lacking tests for this behavior, though.

Reply:
OK, I see what you're getting at. You shouldn't put the canAddBlock in an 'else' branch though, although at the moment the logic is consistent it may not always be. canAddBlock should be check every time someone tries to add a block, without exception. Also, as a user I would probably call canAddBlock() before calling addBlock() everytime - at least that would seem the sensible thing to do, and at the moment canAddBlock contains the block already added + pruned immediately checks that you're talking about - so I would never get the false return value.

Reply:
For now, moved the check at the beginning of the function in 44a72c72794afd9765aeadb305a8bbc959293140; much cleaner looking, indeed.

Still thinking if the semantic can be simplified a bit.